Zend/zend_alloc.c

```diff
@@ -2101,10 +2133,13 @@ static zend_mm_heap *zend_mm_init(void)
 	heap->use_custom_heap = ZEND_MM_CUSTOM_HEAP_NONE;
 #endif
 #if ZEND_MM_STORAGE
-	heap->storage = NULL;
+	ZEND_MM_RO_HEAP(heap)->storage = NULL;
 #endif
 	heap->huge_list = NULL;
 	heap->pid = getpid();
+
+	mprotect(ZEND_MM_RO_HEAP(heap), zend_mm_get_page_size(), PROT_READ);
```

Comment:
Please wrap calls to `mprotect()` in a wrapper function like we do with `zend_mm_mmap()`, so that we can support platforms without `mprotect()`.

We should also check that the wrapper succeeds where we call it (with `ZEND_MM_CHECK()`).